README

Net::Pcap does the job, but it's confusing and difficult to operate.  This
module solves that problem in a relatively general way.

I wanted a way to build libpcap filters really quickly.  It shouldn't take two
hours to crank one out, particularly since I imagine they all kinda look the
same.  So, I wrote this to simplify writing sniffers really quickly.

use strict;
use warnings;
use Net::Pcap::Easy;

# all arguments to new are optoinal
my $npe = Net::Pcap::Easy->new(
    dev              => "lo",
    filter           => "host 127.0.0.1 and icmp",
    packets_per_loop => 10,
    bytes_to_capture => 1024,
    timeout_in_ms    => 0, # 0ms means forever
    promiscuous      => 0, # true or false

    icmp_callback => sub {
        my ($npe, $ether, $ip, $icmp) = @_;

        print "ICMP: $ether->{src_mac}:$ip->{src_ip} -> $ether->{dest_mac}:$ip->{dest_ip}\n";
    },
);

1 while $npe->loop;


Changes
1.4210: Mon Jan 05 2015
   - add a close() method and support early shutdown (useful in SIG{BLAH} = sub { $npe->close }
   - found annoying memory leak from 2012 or whenever I found it (and then forgot about it)

1.4209: Mon Nov 17 2014
   - timeouts will rarely (if ever) work — or do anything like what the user
     expects to happen.  Changed docs and added forks.pm example

1.4207: Tue Jun 07 2011
   - added ->stats()

1.4205: Fri Jul 30 2010
   - accepted a patch from Jerry Litteer
   - added tcp to the synopsis and examples/synopsis.pl

1.4200: Wed Jul 14 2010
   - got a patch from Daniel Roethlisberger regarding packets
     without Ethernet encapsulation
   - wrote some tests for it
   - noted that we insert this fake encapsulation.

1.4105: Fri Dec 25 2009
   - added docs regarding the raw collector
   - added the examples from the pod to an examples dir

1.4100: Fri Dec 18 2009
   - support for fast callbacks, NetPacket::decode() is kinda
     slow since it involves a few function calls.  Per packet,
     this adds up in a hurry.

1.4000: Thu Jun 25 16:00:00 EDT 2009
   - made {pcap} and {netmask} into accessor functions (get_pcap,
     etc) and altered pods
   - got the pods to pass t/pod.t
   - t/pod_coverage.t passes
   - got t/critic.t to pass
   - renamed and moved a few things.
   - to avoid the fact that loop() returns true forever at the
     end of a packetfile, I've made loop() instead return the
     number of processed packets (in cases where there was no
     error) -- and undef or () when there was an error.
   - added the new dev=>"file:lo.data" format to the pod
   - added the dev=>"file:lo.data" test (t/31_)
   - added the dev=>"file:eth0.data" test (t/31_)

1.325: Thu Jun 25 08:41:37 EDT 2009
   - adding patch from andk
   - made sure this works under 5.6.1

1.325: Fri Mar 27 18:28:59 EDT 2009
   - gah, bug in the Makefile.PL, noticed it when I installed on
     my new server.

1.322: Thu Mar 26 06:36:05 EDT 2009
   - There were a bunch of documentation bugs.
   - I fixed some lousy Makefile.PL code

1.321: Wed Mar 25 20:54:13 EDT 2009
   - added chorny's META.yml stuff

1.32: Wed Aug 13 13:49:19 EDT 2008
    - woops, 'skip' device fails tests...

1.31: Wed Aug 13 07:21:36 EDT 2008
    - I dislike module::build, removed

1.3: Wed Aug  6 07:00:43 EDT 2008
    - Awesome.  t/07 had a naked (non eval"") use WWW::Mechanize,
      even though it didn't use it.  So, CPAN testers rock.

1.2: Tue Jul 29 14:42:08 EDT 2008
    - typo in the pod
    - bug in the module
    - I'm not going to include the Makefile.PL anymore, it
      causes problems.

1.1: Mon Jul 28 06:42:14 EDT 2008
    - error in the Build.PL, must set default for prompt() so the
      smokers don't get an error.

1.0: Wed Jul 23 18:41:34 EDT 2008
    - built a tcp test [05], more trouble than you'd think... requires
      root, my default gateway is on eth2, etc...
    - added an icmp test [07]
    - added a tcp+icmp test [10]
      - same test one def filter [11]
      - same test one ipv4 filter [11]
    - added a tcp+icmp using two pcaps test [15]
      - same, but interleaved
      - same, but serialized
    - documented this mess

1.0: Wed Jul 23 09:45:24 EDT 2008
    - I found this stuff pretty hard to do the first time,
      knowing I'd want to do it all again without having to
      re-learn it all, I wrote this.


sudo_and_test.sh
#!/bin/bash
# vi:syntax=gentoo-init-d:

set -e

unset TEST_DEVICE
export PERL_MM_USE_DEFAULT=1
perl Makefile.PL
make


def_route_dev=$(ip route | grep ^default | head -n 1 | perl -nE 'say $1 if m/dev\s+(\S+)/')
echo ${DEV:-$def_route_dev} > device
echo -n "using device="; cat device
sudo prove
echo -n > device



Easy.pm

package Net::Pcap::Easy;

use strict;
no warnings;

use Carp;
use Socket;
use Net::Pcap;
use Net::Netmask;
use NetPacket::Ethernet qw(:types);
use NetPacket::IP qw(:protos);
use NetPacket::ARP qw(:opcodes);
use NetPacket::TCP;
use NetPacket::UDP;
use NetPacket::IGMP;
use NetPacket::ICMP qw(:types);

our $VERSION     = "1.4210";
our $MIN_SNAPLEN = 256;
our $DEFAULT_PPL = 32;

my %KNOWN_CALLBACKS = (map {($_=>1)} qw(
    appletalk_callback arp_callback arpreply_callback arpreq_callback default_callback icmp_callback
    icmpechoreply_callback icmpunreach_callback icmpsourcequench_callback icmpredirect_callback
    icmpecho_callback icmprouteradvert_callback icmproutersolicit_callback icmptimxceed_callback
    icmpparamprob_callback icmptstamp_callback icmptstampreply_callback icmpireq_callback
    icmpireqreply_callback igmp_callback ipv4_callback ipv6_callback ppp_callback rarpreply_callback
    rarpreq_callback snmp_callback tcp_callback udp_callback
));

sub DESTROY {
    my $this = shift;
    $this->close;
    %$this = ();
    return;
}

sub close {
    my $this = shift;

    my $p = delete $this->{pcap};
    Net::Pcap::close($p) if $p;

    return;
}

sub is_local {
    my $this = shift;
    my $nm = $this->cidr;

    my $r = eval { $nm->contains( @_ ) }; croak $@ if $@;
    return $r;
}

sub new {
    my $class = shift;
    my $this = bless { @_ }, $class;

    my $err;
    my $pcap;
    unless ($this->{pcap}) {
        my $dev = $this->{dev};

        if( $dev =~ s/^file:// ) {
            $pcap = $this->{pcap} = 
                Net::Pcap::open_offline($dev, \$err)
                    or die "error opening offline pcap file: $err";

        } else {
            unless( $dev ) {
                $dev = $this->{dev} = Net::Pcap::lookupdev(\$err);
                croak "ERROR while trying to find a device: $err" unless $dev;
            }

            my ($network, $netmask);
            if (Net::Pcap::lookupnet($dev, \$network, \$netmask, \$err)) {
                croak "ERROR finding net and netmask for $dev: $err";

            } else {
                $this->{network} = $network;
                $this->{netmask} = $netmask;
            }

            my $ppl = $this->{packets_per_loop};
               $ppl = $this->{packets_per_loop} = $DEFAULT_PPL unless defined $ppl and $ppl > 0;

            my $ttl = $this->{timeout_in_ms} || 250;
               $ttl = 250 if $ttl < 0;

            my $snaplen = $this->{bytes_to_capture} || 1024;
               $snaplen = $MIN_SNAPLEN unless $snaplen >= 256;

            $pcap = $this->{pcap} = Net::Pcap::open_live($dev, $snaplen, $this->{promiscuous}, $ttl, \$err);

            croak "ERROR opening pacp session: $err" if $err or not $pcap;
        }

        for my $f (grep {m/_callback$/} keys %$this) {
            croak "the $f option does not point to a CODE ref" unless ref($this->{$f}) eq "CODE";
            warn  "the $f option is not a known callback and will never get called" unless $KNOWN_CALLBACKS{$f};
        }
    }

    if( my $f = $this->{filter} ) {
        my $filter;
        Net::Pcap::compile( $pcap, \$filter, $f, 1, $this->{netmask} ) && croak 'ERROR compiling pcap filter';
        Net::Pcap::setfilter( $pcap, $filter ) && die 'ERROR Applying pcap filter';
    }

    $this->{_mcb} = sub {
        my ($linktype, $header, $packet) = @_;

        # For non-ethernet data link types, construct a
        # fake ethernet header from the data available.
        my ($ether, $type);
        if ($linktype == Net::Pcap::DLT_EN10MB) {
            $ether = NetPacket::Ethernet->decode($packet);
            $type = $ether->{type};

        } elsif ($linktype == Net::Pcap::DLT_LINUX_SLL) {
            use bytes;
            $type = unpack("n", substr($packet, 2+2+2+8, 2));
            $ether = NetPacket::Ethernet->decode(
                    pack("h24 n", "0" x 24, $type) . substr($packet, 16));
            no bytes;

        } else {
            die "ERROR Unhandled data link type: " .
                Net::Pcap::datalink_val_to_name($linktype);
        }
        $this->{_pp} ++;

        my $cb;

        return $this->_ipv4( $ether, NetPacket::IP  -> decode($ether->{data}),  $header) if $type == ETH_TYPE_IP;
        return $this->_arp(  $ether, NetPacket::ARP -> decode($ether->{data}),  $header) if $type == ETH_TYPE_ARP;
        
        return $cb->($this, $ether,  $header) if $type == ETH_TYPE_IPv6      and $cb = $this->{ipv6_callback};
        return $cb->($this, $ether,  $header) if $type == ETH_TYPE_SNMP      and $cb = $this->{snmp_callback};
        return $cb->($this, $ether,  $header) if $type == ETH_TYPE_PPP       and $cb = $this->{ppp_callback};
        return $cb->($this, $ether,  $header) if $type == ETH_TYPE_APPLETALK and $cb = $this->{appletalk_callback};

        return $cb->($this, $ether,  $header) if $cb = $this->{default_callback};
    };

    return $this;
}

sub _icmp {
    my ($this, $ether, $ip, $icmp, $header) = @_;

    my $cb;
    my $type = $icmp->{type};

    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_ECHOREPLY     and $cb = $this->{icmpechoreply_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_UNREACH       and $cb = $this->{icmpunreach_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_SOURCEQUENCH  and $cb = $this->{icmpsourcequench_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_REDIRECT      and $cb = $this->{icmpredirect_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_ECHO          and $cb = $this->{icmpecho_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_ROUTERADVERT  and $cb = $this->{icmprouteradvert_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_ROUTERSOLICIT and $cb = $this->{icmproutersolicit_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_TIMXCEED      and $cb = $this->{icmptimxceed_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_PARAMPROB     and $cb = $this->{icmpparamprob_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_TSTAMP        and $cb = $this->{icmptstamp_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_TSTAMPREPLY   and $cb = $this->{icmptstampreply_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_IREQ          and $cb = $this->{icmpireq_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_IREQREPLY     and $cb = $this->{icmpireqreply_callback};

    # NOTE: MASKREQ is exported as MASREQ ... grrz: http://rt.cpan.org/Ticket/Display.html?id=37931
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == NetPacket::ICMP::ICMP_MASKREQ() and $cb = $this->{icmpmaskreq_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $type == ICMP_MASKREPLY     and $cb = $this->{icmpmaskreply_callback};

    return $cb->($this, $ether, $ip, $icmp, $header ) if $cb = $this->{icmp_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $cb = $this->{ipv4_callback};
    return $cb->($this, $ether, $ip, $icmp, $header ) if $cb = $this->{default_callback};

    return;
}

sub _ipv4 {
    my ($this, $ether, $ip, $header) = @_;

    my $cb;
    my $proto = $ip->{proto};

    # NOTE: this could probably be made slightly more efficient and less repeatative.

    return $cb->($this, $ether, $ip, NetPacket::TCP  -> decode($ip->{data}), $header) if $proto == IP_PROTO_TCP  and $cb = $this->{tcp_callback};
    return $cb->($this, $ether, $ip, NetPacket::UDP  -> decode($ip->{data}), $header) if $proto == IP_PROTO_UDP  and $cb = $this->{udp_callback};
    return $this->_icmp($ether,$ip,  NetPacket::ICMP -> decode($ip->{data}), $header) if $proto == IP_PROTO_ICMP;
    return $cb->($this, $ether, $ip, NetPacket::IGMP -> decode($ip->{data}), $header) if $proto == IP_PROTO_IGMP and $cb = $this->{igmp_callback};

    my $spo;
       $spo = NetPacket::TCP  -> decode($ip->{data}) if $proto == IP_PROTO_TCP;
       $spo = NetPacket::UDP  -> decode($ip->{data}) if $proto == IP_PROTO_UDP;
       $spo = NetPacket::IGMP -> decode($ip->{data}) if $proto == IP_PROTO_IGMP;

    return $cb->($this, $ether, $ip, $spo, $header) if $cb = $this->{ipv4_callback};
    return $cb->($this, $ether, $ip, $spo, $header) if $cb = $this->{default_callback};

    return;
}

sub _arp {
    my ($this, $ether, $arp, $header) = @_;

    my $cb;
    my $op = $arp->{opcode};

    return $cb->($this, $ether, $arp, $header) if $op ==  ARP_OPCODE_REQUEST and $cb = $this->{arpreq_callback};
    return $cb->($this, $ether, $arp, $header) if $op ==  ARP_OPCODE_REPLY   and $cb = $this->{arpreply_callback};
    return $cb->($this, $ether, $arp, $header) if $op == RARP_OPCODE_REQUEST and $cb = $this->{rarpreq_callback};
    return $cb->($this, $ether, $arp, $header) if $op == RARP_OPCODE_REPLY   and $cb = $this->{rarpreply_callback};

    return $cb->($this, $ether, $arp, $header) if $cb = $this->{arp_callback};
    return $cb->($this, $ether, $arp, $header) if $cb = $this->{default_callback};

    return;
}

sub loop {
    my $this = shift;
    my $cb   = shift || $this->{_mcb};

    return unless exists $this->{pcap}; # in case we close early

    my $ret = Net::Pcap::loop($this->{pcap}, $this->{packets_per_loop}, $cb, Net::Pcap::datalink($this->{pcap}));

    return unless $ret == 0;
    return (delete $this->{_pp}) || 0; # return the number of processed packets.
}

sub pcap        { return $_[0]->{pcap} }
sub raw_network { return $_[0]->{network} }
sub raw_netmask { return $_[0]->{netmask} }
sub dev         { return $_[0]->{dev} }

sub network {
    my $this = shift;

    return Socket::inet_ntoa(scalar reverse pack("l", $this->{network}));
}

sub netmask {
    my $this = shift;

    return Socket::inet_ntoa(scalar reverse pack("l", $this->{netmask}));
}

sub cidr {
    my $this = shift;
    my $nm = $this->{nm};
       $nm = $this->{nm} = Net::Netmask->new($this->network . "/" . $this->netmask) unless $this->{nm};

    return $nm;
}

sub stats {
    my $this = shift;

    return unless exists $this->{pcap}; # in case we close early

    my %stats;
    Net::Pcap::pcap_stats($this->{pcap}, \%stats);
    $stats{ substr $_, 3 } = delete $stats{$_} for keys %stats;

    return wantarray ? %stats : \%stats;
}

1;


Build.PL

use Module::Build;
my $build = Module::Build->new(
    module_name       => 'Net::Pcap::Easy',
    dist_version_from => "Easy.pm",
    license           => 'lgpl',
     pm_files         => { 'Easy.pm'  => 'lib/Net/Pcap/Easy.pm'  },
    pod_files         => { 'Easy.pod' => 'lib/Net/Pcap/Easy.pod'  },

    build_requires    => {
        # NOTE: They have to be root *and* specify a device to run the tests on, 
        # if they're going to get that interactive they're probably not CPAN testers anyway.
        #
        # 'WWW::Mechanize' => 0,
        # 'Net::Ping'      => 0,

        'File::Slurp' => 0, # the tests won't compile at all without this though
    },

    requires          => {
        'Net::Pcap'           => 0,
        'NetPacket::Ethernet' => 0,
        'NetPacket::IP'       => 0,
        'NetPacket::TCP'      => 0,
        'NetPacket::UDP'      => 0,
        'NetPacket::ARP'      => 0,
        'NetPacket::ICMP'     => 0,
        'Net::Netmask'        => 0,
        'Socket'              => 0,
    },

    add_to_cleanup => ['device'],
);

if( my $d = ($ENV{TEST_DEVICE} || $build->prompt("network device on which to run tests", "skip")) ) {
    $d = '' if $d eq "skip";

    open my $out, ">device" or die $!;
    print $out $d;
    close $out;
}

$build->create_build_script;



Makefile.PL
# vi:set syntax=perl:

use ExtUtils::MakeMaker;

WriteMakefile(
    'NAME'          => 'Net::Pcap::Easy',
    'VERSION_FROM'  => 'Easy.pm',

    'PREREQ_PM'     => { 
        # NOTE: They have to be root *and* specify a device to run the tests on, 
        # if they're going to get that interactive they're probably not CPAN testers anyway.
        #
        # 'WWW::Mechanize' => 0,
        # 'Net::Ping'      => 0,
        'File::Slurp'         => 0, # the tests won't compile at all without this though

        'Net::Pcap'           => 0,
        'NetPacket::Ethernet' => 0,
        'NetPacket::IP'       => 0,
        'NetPacket::TCP'      => 0,
        'NetPacket::UDP'      => 0,
        'NetPacket::ARP'      => 0,
        'NetPacket::ICMP'     => 0,
        'Net::Netmask'        => 0,
        'Socket'              => 0,
    },

    ($ExtUtils::MakeMaker::VERSION ge '6.48'? 
        (MIN_PERL_VERSION => 5.008,
            META_MERGE => {
                keywords => ['pcap', 'sniffer'],
                resources=> {
                    repository => 'http://github.com/jettero/net--pcap--easy/',
                },
            },

        LICENSE	=> 'LGPL',
    ) : ()),

    clean => { FILES => 'device' },
);

my $dev = $ENV{TEST_DEVICE}
       || prompt( "network device on which to run tests (e.g., eth0, must be root): ", "skip");
   $dev = "" if $dev =~ m/skip/;

open my $dh, ">", "device" or die $!;
print $dh $dev;
close $dh;


examples/raw.pl
#!/usr/bin/perl

use strict;
use warnings;
use Net::Pcap::Easy;

my $ppl    = 8192;
my $snap   = 1024;
my $local  = "192.168.0.0/16";
my $lbytes = "\xc0\xa8";
my $upload = 0; # true or false

my $npe = Net::Pcap::Easy->new(
   dev              => ($ARGV[0] || "eth0"),
   filter           => join(" or ", ($upload ? ("(src net $local and not dst net $local)") : ()),
                                                "(dst net $local and not src net $local)" ),
   packets_per_loop => $ppl,
   bytes_to_capture => $snap,

   timeout_in_ms    => 0, # 0ms means forever
   promiscuous      => 0, # true or false
);

sub _do_things {
    my ($cust, $dir, $prot, $port);

    local $" = ", ";
    print "_do_things(@_)\n";
}

1 while defined
# NOTE: defined, since loop returns 0 and undef on error

$npe->loop( sub {
    my ($user_data, $header, $raw_bytes) = @_;

    # $user_data is literally "user data"

    # $header is like this:
    # { caplen => 96, len => 98, tv_sec => 1245963414, tv_usec => 508250 },

    # print unpack("H*", $raw_bytes), "\n";

    my $packet = $_[-1]; # this is the same as $raw_bytes, but I prefer
                         # the word packet and rarely examin the other
                         # elements of @_

    # Calculating precisely what you need is quite a bit faster than
    # decoding the whole packet -- although it's not exactly "Easy."  Say
    # we're interested in IPv4 TCP and UDP packets only... The following
    # assumes *Ethernet* and we don't check it!

    my $l3protlen = ord substr $packet, 14, 1; # the protocol and length
    my $l3prot    = $l3protlen & 0xf0 >> 2; # the protocol part

    return unless $l3prot == 4; # return unless IPv4

    my $l4prot = ord substr $packet, 23, 1; # the L4protocol

    # return unless it's udp or tcp
    return unless $l4prot == 6 or $l4prot == 17;

    my $l3hlen= ($l3protlen & 0x0f) * 4; # number of 32bit words
    my $l4 = 14 + $l3hlen; # the layer 4 data starts here

    # my $src_ip = substr $packet, 26, 4; # these are netowrk order packed
    # my $dst_ip = substr $packet, 30, 4; # but they're pretty easy to convert

    # The ord of the individual bytes of an IP are what we usually
    # see... join(".", map { ord $_} split "", "\xc0\xa8\x01\x01")
    # gives 192.168.1.1

    # Here, I'm only interestd in local network downloads

    if( substr($packet, 30, 2) eq $lbytes ) { # 192.168.0.0/16x.x
        # download direction

        my $cust = ord(substr($packet, 32,1)) .'.'. ord(substr($packet, 33,1));
        my $port = unpack 'n', substr $packet, $l4+0, 2; # src port

        # On my network, in this sense, the last two bytes idenfity a "customer"
        # the source port (on a download) is the port the protocol operates on.
        # 80 for http, 110 for pop3, etc.  Unpack('n', blarg) gives the
        # network order 2-byte port number as a Perl number.

        _do_things($cust, dn => $l4prot == 6 ? "tcp" : "udp", $port);

    } elsif( $upload ) {
        # upload direction

        my $cust = ord(substr($packet, 28,1)) .'.'. ord(substr($packet, 29,1));
        my $port = unpack 'n', substr $packet, $l4+2, 2; # dst port

        # In the upload direction, the protocol port is the dst port.

        _do_things($cust, up => $l4prot == 6 ? "tcp" : "udp", $port);
    }

});
